raster-foundry/granary
# Support continuous deployment to AWS

## Overview

Make the following changes in order to support continuous deployment to AWS:

- Replace CircleCI with Jenkins (Read: Notes)
- Add Terraform resources to support an ECS Fargate-based service

## Notes

The output of `terraform plan` contained secrets that are passed to the ECS task definition, and these were exposed on CircleCI during every deployment. I evaluated the following paths before deciding on jumping ship to Jenkins:

- Split the deployment out into a Jenkins job. I didn't want to have the build split across multiple CI platforms. Also, there [isn't](https://support.circleci.com/hc/en-us/articles/115014372807-IP-Address-ranges-for-whitelisting-Do-you-have-static-IP-addresses-available-) a way to whitelist CircleCI IP address ranges to allow the upstream pipeline to trigger the Jenkins job. It looked like I'd have to have both jobs listen to Granary via a webhook, and this seemed inelegant, as Jenkins would cycle the ECS task definition before there is a container image present. 

- Store secrets with Secrets Manager. This would add an additional layer of complexity between deciding what belongs in `terraform.tfvars` vs. Secrets Manager. It's also a net new service to us, and I don't think this is the place to be experimenting with something like that. 

See: https://docs.aws.amazon.com/AmazonECS/latest/developerguide/specifying-sensitive-data.html


- Pipe `terraform plan` to `/dev/null`. 😬

## Testing Instructions

```bash
$ http -F granary.rasterfoundry.com/api/hello/world
HTTP/1.1 200 OK
Connection: keep-alive
Content-Length: 26
Content-Type: application/json
Date: Fri, 08 Nov 2019 19:37:19 GMT

{
    "message": "Hello, world"
}
```

```bash
$ http -F granary.rasterfoundry.com/api/models
HTTP/1.1 200 OK
Connection: keep-alive
Content-Length: 2
Content-Type: application/json
Date: Tue, 19 Nov 2019 16:18:07 GMT

[]
```

See Jenkins build: http://jenkins.staging.rasterfoundry.com/job/Raster%20Foundry/job/granary/job/jrb%252Fecs/4/

See CloudWatch logs: https://console.aws.amazon.com/cloudwatch/home?region=us-east-1#logEventViewer:group=logGranaryAPI

Resolves #9
